### PR TITLE
filter data-workspace projects by ext

### PR DESCRIPTION
**Pull request by kisantia**
This updates the dataworkspace.d.ts api to filter the projects by the selected project extension so that each extension doesn't have to implement the filtering themselves, as requested by Alex

### CODE REVIEW

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 1]
```diff
@@ -155,8 +155,15 @@ export class WorkspaceService implements IWorkspaceService {
...
 	}
 
-	getProjectsInWorkspace(): vscode.Uri[] {
-		return vscode.workspace.workspaceFile ? this.getWorkspaceConfigurationValue<string[]>(ProjectsConfigurationName).map(project => this.toUri(project)) : [];
+	getProjectsInWorkspace(ext?: string): vscode.Uri[] {
+		let projects = vscode.workspace.workspaceFile ? this.getWorkspaceConfigurationValue<string[]>(ProjectsConfigurationName).map(project => this.toUri(project)) : [];
+
+		// filter by specified extension
+		if (ext) {
+			projects.filter(p => path.extname(p.fsPath) === ext);
```
Comment (alanrenmsft) [inline extensions/data-workspace/src/services/workspaceService.ts:163]: 1. we should handle the casing as well as whether there is a leading '.'. `sqlProj`, and `.sqlproj` should both work
2. you have to assign this to the projects variable, filter method won't update the original array

File: extensions/data-workspace/src/dataworkspace.d.ts [hunk 2]
```diff
@@ -15,8 +15,9 @@ declare module 'dataworkspace' {
 	export interface IExtension {
 		/**
 		 * Returns all the projects in the workspace
+		 * @param ext extension type to filter on
```
Comment (alanrenmsft) [inline extensions/data-workspace/src/dataworkspace.d.ts:18]: a bit more description here?

**Review by alanrenmsft — CHANGES_REQUESTED** [review]
left a couple comments

File: extensions/data-workspace/src/services/workspaceService.ts (hunk 1 above)
Comment (kisantia) [reply extensions/data-workspace/src/services/workspaceService.ts]: oops thanks for catching

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 3]
```diff
@@ -155,8 +155,15 @@ export class WorkspaceService implements IWorkspaceService {
...
 	}
 
-	getProjectsInWorkspace(): vscode.Uri[] {
-		return vscode.workspace.workspaceFile ? this.getWorkspaceConfigurationValue<string[]>(ProjectsConfigurationName).map(project => this.toUri(project)) : [];
+	getProjectsInWorkspace(ext?: string): vscode.Uri[] {
+		let projects = vscode.workspace.workspaceFile ? this.getWorkspaceConfigurationValue<string[]>(ProjectsConfigurationName).map(project => this.toUri(project)) : [];
+
+		// filter by specified extension
+		if (ext) {
+			projects = projects.filter(p => p.fsPath.endsWith(ext));
```
Comment (alanrenmsft) [inline extensions/data-workspace/src/services/workspaceService.ts:163]: p.fsPath.toLowerCase().endsWith(ext.toLowerCase())

Comment (Benjin) [reply extensions/data-workspace/src/services/workspaceService.ts:163]: the OS community at large should just ban all non-lowercased paths. :D

**Review by Benjin — APPROVED** [review]
+1 to Alan's comment, otherwise good

**Review by alanrenmsft — APPROVED** [review]

File: extensions/data-workspace/src/services/workspaceService.ts (hunk 3 above)
Comment (nahk-ivanov) [reply extensions/data-workspace/src/services/workspaceService.ts]: Just an FYI, it should be `toUpperCase`, not `toLowerCase`.

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1308

**Pull request merged**